ddtrace/internal/globaltracer.go

```
@@ -0,0 +1,41 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2025 Datadog, Inc.
+
+package internal
+
+import "sync/atomic"
+
+var (
+	// globalTracer stores the current tracer as *ddtrace/tracer.Tracer (pointer to interface). The
+	// atomic.Value type requires types to be consistent, which requires using the same type for the
+	// stored value.
+	globalTracer atomic.Value
+)
+
+// tracerLike is an interface to restrict the types that can be stored in `globalTracer`.
+// This interface doesn't leak to the users. We are leveraging the type system to generate
+// the functions below for `tracer.Tracer` without creating an import cycle.
+type tracerLike interface {
+	Flush()
+	Stop()
+}
+
+// SetGlobalTracer sets the global tracer to t.
+// It is the responsibility of the caller to ensure that the value is `tracer.Tracer`.
+func SetGlobalTracer[T tracerLike](t T) {
+	old := globalTracer.Swap(&t)
+	if old == nil {
+		return
+	}
+	oldTracer := *old.(*T)
+	oldTracer.Stop()
+}
+
+// GetGlobalTracer returns the current global tracer.
+// It is the responsability of the caller to ensure that calling code uses `tracer.Tracer`
+// as generic type.
+func GetGlobalTracer[T tracerLike]() T {
+	return *globalTracer.Load().(*T)
```

Should we also be checking for a nil `Load()` return here?

---

In the context of our tracer, this shouldn't happen. The `init` function ensures `globalTracer` is set. If somebody misuses the API and sets a `nil` value, which is possible, the check would make sense.

The problem with the check in this PR is that... What should we return? Returning `nil` would lead to panics when the tracer API is used. Another way would be to set a `NoopTracer`, but that would create an import cycle between `internal` and `tracer`.

I think the best way would be to panic if `SetGlobalTracer` is called with a `nil` value. A `nil` tracer could be considered a bad state that is not recoverable. WDYT? cc @kakkoyun

---

@darccio I have no strong opinion about panicking in the `SetGlobalTracer` when a `nil` value is passed but we could also do nothing when called with a nil instead of panicking.

---

@genesor My fear by doing nothing is that it might go unnoticed. Another way would be to log that a `nil` tracer has been passed and we avoided setting it. WDYT?

---

@darccio I'd go for the log, that's less "radical" than a panic IMO

---

@genesor Now I realized that this being an `internal` package, it doesn't have access to our logger. If we use `log` as provided by the stdlib, it would be a deviation from the rest of the tracer.

I'm reconsidering going the panic route.

---

IMO an early panic is acceptable in this path. If user miss uses the API, we can fail fast.

But using the `NoopTracer` and having a warning would be ideal. How was the previous behaviour?

---

Panic route took in 02063e7f387ebd317f4df34ba3a706317197ed8e

---

The explanation: in `v1` it was also possible to set a `nil` tracer. It causes the code to panic once a span is started. I feel it's better to fail on initialization than during the execution.